Reflecting particle boundary conditions

gsgall opened the pull request:
This PR implements reflective boundary conditions for particles addressing issue #39.

gsgall committed:
Initial particle reflection boundary condition addressing issue #39

- Reflecting particles requires updating the velocity stored in RayData which the ReflectRayBC has no knowledge of
- This BC ensures that the ray data and ray direction are consistent

gsgall committed:
Adding tests for boundary conditions in all dimensions

- each test ensurs components in which rays are not moving are not modified

gsgall committed:
Updating the documentation for the boundary condition

gsgall commented:
@simopier @cticenhour @loganharbour When you guys get a chance could you take a look at this PR?

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3061445/), step Sync to remote on 5c8a5d5 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/salamander/docs/PRs/88/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3061450/), step Generate coverage on 5c8a5d5 wanted to post the following:


## Coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>3bf1d3</th>
      <th colspan=4>#88 5c8a5d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>93.71%</td>
      <td>93.90%</td>
      <td>+0.18%</td>
      <td>95.83%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>626</td>
      <td>646</td>
      <td>+20</td>
      <td>23</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>42</td>
      <td>42</td>
      <td>-</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/salamander/docs/PRs/88/coverage_diff)

[Full coverage report](https://mooseframework.inl.gov/salamander/docs/PRs/88/coverage)



This comment will be updated on new commits.

simopier commented on doc/content/source/raybcs/ReflectParticleBC.md:
```diff
@@ -0,0 +1,16 @@
+
+!syntax description /RayBCs/ReflectParticleBC
+
+Since particle velocity data is stored in RayData it will not be updated by the [ReflectRayBC.md]. This boundary condition performs the same operation as the [ReflectRayBC.md] and then updates the particles velocity data ensuring it is consistent with the particles direction after reflection.
```
```suggestion
Since particle velocity data is stored in `RayData`, it will not be updated by [ReflectRayBC.md]. 
This boundary condition performs the same operation as [ReflectRayBC.md] and then updates the particles velocity data ensuring it is consistent with the particles direction after reflection.
```

simopier commented on doc/content/source/raybcs/ReflectParticleBC.md:
```diff
@@ -0,0 +1,16 @@
+
+!syntax description /RayBCs/ReflectParticleBC
+
+Since particle velocity data is stored in RayData it will not be updated by the [ReflectRayBC.md]. This boundary condition performs the same operation as the [ReflectRayBC.md] and then updates the particles velocity data ensuring it is consistent with the particles direction after reflection.
```
Please elaborate a bit more on why things stored in RayData would not be updated by ReflectRayBC.
Maybe explain how a particle is different from a ray, and why it needs more.

gsgall replied:
@simopier I added some additional context to this page please let me know if you would like any additional information on this page.

simopier commented on include/raybcs/ReflectParticleBC.h:
```diff
@@ -0,0 +1,37 @@
+  const unsigned int _mesh_dimension;
+  /// the ray data indicies for the velocities stored on the ray
+  const std::vector<RayDataIndex> _velocity_indicies;
+  /// point used to store the components of the velocity which are needed for update during reflection
```
```suggestion
  /// point used to store the components of the velocity which are needed during reflection
```

simopier commented on src/raybcs/ReflectParticleBC.C:
```diff
@@ -0,0 +1,57 @@
+{
+  auto params = ReflectRayBC::validParams();
+  params.addClassDescription(
+      "Relfective boundary condition for particles that ensures velocity data is consistent with "
```
```suggestion
      "Reflective boundary condition for particles that ensures velocity data is consistent with "
```

simopier commented on src/raybcs/ReflectParticleBC.C:
```diff
@@ -0,0 +1,57 @@
+  auto params = ReflectRayBC::validParams();
+  params.addClassDescription(
+      "Relfective boundary condition for particles that ensures velocity data is consistent with "
+      "its direction after relfection.");
```
```suggestion
      "its direction after reflection.");
```

simopier commented on test/tests/raybcs/1d.i:
```diff
@@ -0,0 +1,22 @@
+  []
+[]
+
+
```
```suggestion
```

simopier commented on test/tests/raybcs/1d.i:
```diff
@@ -0,0 +1,22 @@
+  nx = 3
+  xmax = 1
+[]
+
```
```suggestion
```

simopier commented on test/tests/raybcs/2d.i:
```diff
@@ -0,0 +1,28 @@
+  xmax = 1
+  ymax = 1
+[]
+
```
```suggestion
```

simopier commented on test/tests/raybcs/2d.i:
```diff
@@ -0,0 +1,28 @@
+  []
+[]
+
+
```
```suggestion
```

simopier commented on test/tests/raybcs/3d.i:
```diff
@@ -0,0 +1,32 @@
+  ymax = 1
+  zmax = 1
+[]
+
```
```suggestion
```

simopier commented on test/tests/raybcs/3d.i:
```diff
@@ -0,0 +1,32 @@
+  []
+[]
+
+
```
```suggestion
```

simopier commented on test/tests/raybcs/reflection_base.i:
```diff
@@ -0,0 +1,54 @@
+    execute_on = TIMESTEP_END
+  []
+[]
+
```
```suggestion
```

simopier commented on test/tests/raybcs/tests:
```diff
@@ -0,0 +1,31 @@
+[Tests]
+  design = 'ReflectParticleBC.md'
+  issues = '#25'
```
```suggestion
  issues = '#39'
```

simopier reviewed (changes requested):
Looks good to me overall, I just have some minor comments for you to address.

gsgall committed:
Apply suggestions from code review

Co-authored-by: Pierre-Clement Simon <pierreclement.simon@gmail.com>

gsgall committed:
Adding additional information as to why a seperate ReflectParticleBC is
required.

simopier reviewed (approved):
Great improvement to the documentation in the last commit! I'm happy with this PR now, but I also want @cticenhour and @loganharbour to jump in.

cticenhour commented:
I am working today, but not on this unfortunately. Will take a look early into the coming week.

loganharbour commented on test/tests/raybcs/reflection_base.i:
```diff
@@ -0,0 +1,53 @@
+  [null]
+    type = NullRayKernel
+  []
+[]
```
FYI you don't need to have this - you can set `ray_kernel_coverage_check = false` in the study if you don't need ray kernels

loganharbour reviewed (approved):
No concerns from me. My one comment isn't a requirement

moosebuild commented:
Job [Test no optional](https://civet.inl.gov/job/3056702/) on 6969217 : invalidated by @cticenhour

No optional documentation allowed to fail

loganharbour commented on src/raybcs/ReflectParticleBC.C:
```diff
@@ -0,0 +1,57 @@
+    _mesh_dimension(_fe_problem.mesh().dimension()),
+    _velocity_indicies({_study.getRayDataIndex("v_x"),
+                        _study.getRayDataIndex("v_y"),
+                        _study.getRayDataIndex("v_z")})
```
Actually - now that I think about it more - shouldn't we add a method to your base study to get these instead? That way we don't need to actually know these strings (v_x, v_y, v_z) across all of the objects that need them.

Do a `getStudy` with whatever your base study type is, and then implement a:

```
std::vector<RayDataIndex>
YourStudy::getVelocityDataIndices()
{
  const auto mesh_dim = _mesh.dimension();
  std::vector<RayDataIndex> indices(mesh_dim);
  for (const auto dim : make_range(mesh_dim))
     indices[dim] = getRayDataIndex("v_" + dim == 0 ? "x" : (dim == 1 ? "y" : "z"));
  return indices;
}
```

And then you can just call that method in each object that needs the velocities. No need to get `_mesh_dimension`in the objects either, as you can just do an `index_range` on that vector. And then, we also don't allocate extra data when we don't need it (higher dims)

gsgall replied:
This definitely makes sense. I can think of a few other places where this data and indices will be needed as well.

cticenhour commented:
Approved from me as well. @gsgall I will leave it to you whether you want to apply Logan's suggestion now or work on it for another PR. Just let me know.

gsgall committed:
Addressing @loganharbour's comments

- Adding getter method for the velocity indicies used to store particle
  velocity data on rays.

- Also updates another object to use this method for getting the
  indicies as well.

loganharbour commented on src/raybcs/ReflectParticleBC.C:
```diff
@@ -0,0 +1,54 @@
+
+ReflectParticleBC::ReflectParticleBC(const InputParameters & params)
+  : ReflectRayBC(params),
+    _velocity_indicies(dynamic_cast<PICStudyBase *>(&_study)->getVelocityIndicies(false))
```
```suggestion
    _velocity_indicies(getStudy<PICStudyBase>().getVelocityIndicies(false))
```

`dynamic_cast` will fail if the study isn't of the right type. `getStudy()` will give a useful error.

loganharbour commented on src/userobjects/PICStudyBase.C:
```diff
@@ -149,3 +151,15 @@ PICStudyBase::setInitialParticleData(std::shared_ptr<Ray> & ray, const InitialPa
+const std::vector<RayDataIndex>
+PICStudyBase::getVelocityIndicies(const bool all_components) const
+{
+  unsigned int vel_dim = all_components ? 3 : _mesh.dimension();
```
```suggestion
  const unsigned int vel_dim = all_components ? 3 : _mesh.dimension();
```

loganharbour commented on src/raybcs/ReflectParticleBC.C:
```diff
@@ -0,0 +1,54 @@
+  for (const auto i : make_range(_velocity_indicies.size()))
+  {
+    currentRay()->data(_velocity_indicies[i]) = _temporary_velocity(i);
+  }
```
```suggestion
  for (const auto i : index_range(_velocity_indicies))
    currentRay()->data(_velocity_indicies[i]) = _temporary_velocity(i);
```

loganharbour commented on src/raybcs/ReflectParticleBC.C:
```diff
@@ -0,0 +1,54 @@
+  for (const auto i : make_range(_velocity_indicies.size()))
+  {
+    _temporary_velocity(i) = currentRay()->data(_velocity_indicies[i]);
+  }
```
```suggestion
  for (const auto i : index_range(_velocity_indicies))
    _temporary_velocity(i) = currentRay()->data(_velocity_indicies[i]);
```

gsgall committed:
Apply suggestions from code review

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

gsgall committed:
Applying formatting corrections

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3061431/), step Clang format on b0e3068 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/salamander/docs/PRs/88/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/salamander/docs/PRs/88/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3bf1d305282f8c90941a4ae4ed41f3157b9698c5`
